src/lib/mdns/minimal/ActiveResolveAttempts.cpp

```diff
@@ -0,0 +1,177 @@
+    ChipLogProgress(Discovery, "Discovered node without a pending query");
+}
+
+void ActiveResolveAttempts::MarkPending(const PeerId & peerId)
+{
+    // Strategy when picking the peer id to use:
+    //   1 if a matching peer id is already found, use that one
+    //   2 if an 'unused' entry is found, use that
+    //   3 otherwise expire the one with the largest nextRetryDelaySec
+    //     or if equal nextRetryDelaySec, pick the one with the oldest
+    //     queryDueTimeMs
+
+    RetryEntry * entryToUse = &mRetryQueue[0];
+
+    for (size_t i = 1; i < kRetryQueueSize; i++)
+    {
+        if (entryToUse->peerId == peerId)
+        {
+            break; // best match possible
+        }
+
+        RetryEntry * entry = mRetryQueue + i;
+
+        // Rule 1: peer id match always matches
+        if (entry->peerId == peerId)
+        {
+            entryToUse = entry;
+            continue;
+        }
+
+        // Rule 2: select unused entries
+        if ((entryToUse->peerId.GetNodeId() != kUndefinedNodeId) && (entry->peerId.GetNodeId() == kUndefinedNodeId))
+        {
+            entryToUse = entry;
+            continue;
+        }
+        else if (entryToUse->peerId.GetNodeId() == kUndefinedNodeId)
+        {
+            continue;
+        }
+
+        // Rule 2: both choices are used (have a defined node id),
+        // pick the one with most queryDueTimeMs
+        if (entry->nextRetryDelaySec > entryToUse->nextRetryDelaySec)
+        {
+            entryToUse = entry;
+        }
+        else if ((entry->nextRetryDelaySec == entryToUse->nextRetryDelaySec) &&
+                 (entry->queryDueTimeMs < entryToUse->queryDueTimeMs))
+        {
+            entryToUse = entry;
+        }
+    }
+
+    if ((entryToUse->peerId.GetNodeId() != kUndefinedNodeId) && (entryToUse->peerId != peerId))
+    {
+        ChipLogError(Discovery, "Re-using pending resolve entry before reply was received.");
+    }
+
+    entryToUse->peerId            = peerId;
+    entryToUse->queryDueTimeMs    = mClock->GetMonotonicMilliseconds();
+    entryToUse->nextRetryDelaySec = 1;
+}
+
+Optional<uint32_t> ActiveResolveAttempts::GetMsUntilNextExpectedResponse() const
+{
+    Optional<uint32_t> minDelay = Optional<uint32_t>::Missing();
+
+    chip::System::Clock::MonotonicMilliseconds nowMs = mClock->GetMonotonicMilliseconds();
+
+    for (auto & entry : mRetryQueue)
+    {
+        if (entry.peerId.GetNodeId() == kUndefinedNodeId)
+        {
+            continue;
+        }
+
+        if (nowMs >= entry.queryDueTimeMs)
+        {
+            // found an entry that needs processing right now
+            return Optional<uint32_t>::Value(0);
+        }
+
+        uint32_t entryDelay = static_cast<int>(entry.queryDueTimeMs - nowMs);
+        if (!minDelay.HasValue() || (minDelay.Value() > entryDelay))
+        {
+            minDelay.SetValue(entryDelay);
+        }
+    }
+
+    return minDelay;
+}
+
+Optional<PeerId> ActiveResolveAttempts::NextScheduledPeer()
```

Comment:
This method implies that the queue will be sorted on query due time.  Is it?

Reply:
It would probably also be good to have a comment that points out that this method is modifying queryDueTimeMs for the retrieved entry.

Reply:
Updated header comment. Thank you for spotting the ordering issue: this was not supposed to rely on sorting as replies  could be received at any time. I will add a unit test and modify the implementation.

Reply:
Updated the comment regarding sorting a bit  more as well: this method will return any 'this peer requires a resolution now' regardless of sorting.

Waiting for at least `GetMsUntilNextExpectedResponse` will ensure at least one entry is returned  by this method, however actual values may be in any order.